src/interface.jl

```diff
@@ -0,0 +1,655 @@
+```
+cb1 = b1 ∘ b2                  # => Composed.ts == (b2, b1)
+cb2 = compose(b2, b1)          # => Composed.ts == (b2, b1)
+cb1(x) == cb2(x) == b1(b2(x))  # => true
+```
+"""
+struct Composed{A} <: Bijector
+    ts::A
+end
+
+compose(ts::Bijector...) = Composed(ts)
+
+# The transformation of `Composed` applies functions left-to-right
+# but in mathematics we usually go from right-to-left; this reversal ensures that
+# when we use the mathematical composition ∘ we get the expected behavior.
+# TODO: change behavior of `transform` of `Composed`?
+∘(b1::Bijector, b2::Bijector) = compose(b2, b1)
+
+inv(ct::Composed) = Composed(map(inv, reverse(ct.ts)))
+
+# # TODO: should arrays also be using recursive implementation instead?
+function (cb::Composed{<: AbstractArray{<: Bijector}})(x)
+    res = x
+    for b ∈ cb.ts
+        res = b(res)
+    end
+
+    return res
+end
+
+# recursive implementation like this allows type-inference
+_transform(x, b1::Bijector, b2::Bijector) = b2(b1(x))
+_transform(x, b::Bijector, bs::Bijector...) = _transform(b(x), bs...)
+(cb::Composed{<: Tuple})(x) = _transform(x, cb.ts...)
+
+function _logabsdetjac(x, b1::Bijector, b2::Bijector)
+    res = forward(b1, x)
+    return logabsdetjac(b2, res.rv) .+ res.logabsdetjac
+end
+function _logabsdetjac(x, b1::Bijector, bs::Bijector...)
+    res = forward(b1, x)
+    return _logabsdetjac(res.rv, bs...) .+ res.logabsdetjac
+end
+logabsdetjac(cb::Composed, x) = _logabsdetjac(x, cb.ts...)
+
+# recursive implementation of `forward`
+# HACK: we need this one in the case where `length(cb.ts) == 2`
+# in which case forward(...) immediately calls `_forward(::NamedTuple, b::Bijector)`
+function _forward(f::NamedTuple, b::Bijector)
+    y, logjac = forward(b, f.rv)
+    return (rv=y, logabsdetjac=logjac .+ f.logabsdetjac)
+end
+function _forward(f::NamedTuple, b1::Bijector, b2::Bijector)
+    f1 = forward(b1, f.rv)
+    f2 = forward(b2, f1.rv)
+    return (rv=f2.rv, logabsdetjac=f2.logabsdetjac .+ f1.logabsdetjac .+ f.logabsdetjac)
+end
+function _forward(f::NamedTuple, b::Bijector, bs::Bijector...)
+    f1 = forward(b, f.rv)
+    f_ = (rv=f1.rv, logabsdetjac=f1.logabsdetjac .+ f.logabsdetjac)
+    return _forward(f_, bs...)
+end
+_forward(x, b::Bijector, bs::Bijector...) = _forward(forward(b, x), bs...)
+forward(cb::Composed{<: Tuple}, x) = _forward(x, cb.ts...)
+
+function forward(cb::Composed, x)
+    rv, logjac = forward(cb.ts[1], x)
+    
+    for t in cb.ts[2:end]
+        res = forward(t, rv)
+        rv = res.rv
+        logjac = res.logabsdetjac .+ logjac
+    end
+    return (rv=rv, logabsdetjac=logjac)
+end
+
+##############################
+# Example bijector: Identity #
+##############################
+
+struct Identity <: Bijector end
+(::Identity)(x) = x
+(::Inversed{Identity})(y) = y
+
+forward(::Identity, x) = (rv=x, logabsdetjac=zero(x))
+
+logabsdetjac(::Identity, y::T) where T <: Real = zero(T)
```

Comment:
Plz remove spaces between `<:` and consider add `{`,`}` to two sides if the line is not too long.

Reply:
Same for all.

Reply:
Is this actually a part of the style-guide? Couldn't find it anywhere. I mean that `f(a::AbstractArray{<: Real})` should be `f(a::AbstractArray{<:Real})` and so on, i.e. no space with the `<:`. 

It seems a bit, uhmm, "cramped" to me. The whitespace inbetween makes it much easier to quickly see that the `<:` is _not_ a part of the type-name. E.g. `{T <: Real, S <: AbstractFloat}` vs. `{T<:Real, S<:AbstractFloat}`.

Reply:
Interesting that this isn't in the style guide. If someone could add it, that would be fantastic.

As regards the actual issue at hand, I'm personally quite pro- the space-free version. Picking a bit of julia at random as evidence, all of the [cholesky](https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/cholesky.jl) code uses even more aggressively space-free conventions than we do, which I'm also not in favour of -- I really do find it hard to read some of that.

Happy to change to a different convention if there's a consensus around something other than what we're doing at the minute and someone is filling to make the required changes to the codebase. Either way though, this should be documented in the style guide.

Reply:
> Interesting that this isn't in the style guide. If someone could add it, that would be fantastic.
It's not mentioned explicitly as far as I can tell, but there are examples using `{T<:Whatever}`. In those cases the "bad" examples are also using the same convention, so there's not a
```julia
# good
f(a::AbstractArray{<:T}) where {T} = ...

# bad
f(a::AbstractArray{<: T}) where T = ...
```
or something. I could add this example to the "Avoid unnecessary whitespace in these situations" section if you'd like.

> As regards the actual issue at hand, I'm personally quite pro- the space-free version. Picking a bit of julia at random as evidence, all of the cholesky code uses even more aggressively space-free conventions than we do, which I'm also not in favour of -- I really do find it hard to read some of that.

I'm probably missing something but to me they seem to be using the same convention?

Reply:
The good / bad examples would be great.

> I'm probably missing something but to me they seem to be using the same convention?

See [here](https://github.com/JuliaLang/julia/blob/c0478d87df6182932b78738be97cc108f0f93248/stdlib/LinearAlgebra/src/cholesky.jl#L347) for example. There isn't any whitespace, and I wouldn't advocate for this.

```julia
# bad
Union{StridedMatrix,RealHermSymComplexHerm{<:Real,<:StridedMatrix}}

# good
Union{StridedMatrix, RealHermSymComplexHerm{<:Real, <:StridedMatrix}}
```
i.e. always adding spaces after commas.

Reply:
Now I follow; thanks!

Reply:
Looks good! Let's stick to Will's version.

Reply:
"i.e. always adding spaces after commas." is easy to remeber :)